### ISSUE TITLE
Huge Training time for FP16 on Turing/2080 GPU

### ISSUE THREAD

**Issue by gvskalyan** [pytorch/fairseq#464]
I am training en-fr on GeForce RTX 2*2080 ti with nvidia driver version 410.79 (although the latest is 410.93) with CUDA 10 and fp16 flag enabled
With the following command 

python3.6 -m torch.distributed.launch --nproc_per_node 2 train.py data-bin --arch transformer --share-all-embeddings --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 --lr 0.0007 --min-lr 1e-09 --dropout 0.1 --weight-decay 0.0 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 --max-tokens 4096 --update-freq 8 --save-dir /experiments/transformer_base_fr_en --ddp-backend=no_c10d --fp16

But a part of my training log is 
| epoch 001:   0%| | 73/21742 [09:49<49:52:09,  8.29s/it, loss=14.730, nll_loss=14.568, ppl=24291.78, wps=4262, ups=0.1, wpb=60960, bsz=1607, num_updates=69, lr=1.21733e-05, gnorm=3.811, clip=0%, oom=0, loss_scale=8.000, wall=987, train_wall=586]

which says 50 more hours are left, but when fp16 is disabled it takes around 10 hours per epoch on the whole 36 M sentences. Why is there such a huge drop ?
Any flags to add ? Or might be a mismatch between driver version and graphic card ?

Please help..

**Comment by zxqchat** [comment]
Theoretically speaking. this parameters --fp16 can be help to training speed , so you can examine your code and I will test it in my experimental.

**Comment by zxqchat** [comment]
@gvskalyan > do not use --fp16 with P40. Only Volta-level GPUs support TensorCores and have fast FP16 training. Pascal-level cards will be faster with FP32.
_Originally posted by @myleott in https://github.com/pytorch/fairseq/issues/247#issuecomment-412398349_

**Comment by myleott** [comment]
2080ti should have Tensor Cores and support fast FP16, but I don't have access to one of these cards to test with.

Have you confirmed that your CUDA + pytorch installation is setup correctly? Can you take a look at this thread and see if anything helps: https://www.reddit.com/r/nvidia/comments/9ikas2/rtx_2080_machine_learning_performance/

You can also try running `nvprof` and look for the "884" kernels. See slide 12 here: http://on-demand.gputechconf.com/gtc-taiwan/2018/pdf/5-1_Internal%20Speaker_Michael%20Carilli_PDF%20For%20Sharing.pdf

**Comment by AranKomat** [comment]
I noticed that torch.matmul and its variants are very slow if CUDA is not up-to-date, so that's probably the culprit.

**Comment by davidecaroselli** [comment]
Is there any chance that 2080 (let's say Turing arch) will support fairseq FP16 properly? Is this something depending on fairseq implementation, or more low level?

**Comment by gvskalyan** [comment]
Even when the training starts say normal (without fp16 flag) 

| NOTICE: your device may support faster training with --fp16 

There would be this message which suggests it (architecture) is capable of fp16 training, but with the flag it is taking a lot of time!

**Comment by myleott** [comment]
The note about faster training is printed based on the device compatibility returned by torch: https://github.com/pytorch/fairseq/blob/fbd4cef9a575b5f77ca05d4b7c3ad3adb11141ac/fairseq/trainer.py#L109-L110

A little background about FP16 training. Training with FP16/mixed precision typically adds extra overhead compared to pure FP32 training, thus why it may be slower than pure FP32 on older hardware. However, the Volta generation of GPUs (and newer) introduced Tensor Cores, which speeds up FP16 operations up to 5x, and makes the overhead worth it; in practice we see a total speedup of ~3x on V100 cards (we lose a bit because of this extra overhead).

The 2080ti is based on the latest Turing architecture and also has Tensor Cores, but not as many as the V100 (the V100 has 640 Tensor Cores compared to 544 on the 2080ti and 368 on the 2080). I'd still expect to see a decent speedup from FP16 on these cards, but I haven't been able to test this myself.

One thing to keep in mind is that it's important that your setup is up-to-date and configured properly. Can you make sure you're using the latest versions of all the libraries listed here (e.g., CUDA 10.0.130, cuBLAS, CuDNN 7.4.2, etc.): https://docs.nvidia.com/deeplearning/dgx/pytorch-release-notes/rel_19-01.html#rel_19-01. Also make sure you're using the latest version of pytorch compiled for 10.0 toolkit.

**Comment by frankang** [comment]
you may try the fp16 version classification, recommendation, etc. tasks provided by NVIDIA and check if fp16/fp32 results doesn't differ.
[https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch](https://github.com/NVIDIA/DeepLearningExamples/tree/master/PyTorch)

**Comment by gvskalyan** [comment]
As suggested in http://on-demand.gputechconf.com/gtc-taiwan/2018/pdf/5-1_Internal%20Speaker_Michael%20Carilli_PDF%20For%20Sharing.pdf 
my O/p is turing_fp16_s1688gemm_fp16_128x128_ldg8_f2f_tn

And fp16 is working on turing architecture. Earlier it might be due to conflict of nvidia toolkit or driver versions etc., 
As already known training time is lesser . Thank You @myleott @davidecaroselli 

Stat : for training en-vi model on 133k sentences (per epoch)
fp32 - 92 sec
fp16 - 48 sec
